lib/logstash/outputs/tcp.rb

```diff
@@ -137,67 +137,81 @@ def register
-        LogStash::Util.set_thread_name("[#{pipeline_id}]|output|tcp|server_accept")
-        loop do
-          break if @closed.value
-          client_socket = server_socket.accept_nonblock exception: false
-          if client_socket == :wait_readable
-            IO.select [ server_socket ]
-            next
-          end
-          Thread.start(client_socket) do |client_socket|
-            # monkeypatch a 'peer' method onto the socket.
-            client_socket.instance_eval { class << self; include ::LogStash::Util::SocketPeer end }
-            @logger.debug("accepted connection", client: client_socket.peer, server: "#{@host}:#{@port}")
-            client = Client.new(client_socket, @logger)
-            Thread.current[:client] = client
-            LogStash::Util.set_thread_name("[#{pipeline_id}]|output|tcp|client_socket-#{@client_threads.size}")
-            @client_threads << Thread.current
-            client.run unless @closed.value
-          end
+    run_as_server if server?
+    run_as_client unless server?
+  end
+
+  def run_as_server
+    @logger.info("Starting tcp output listener", :address => "#{@host}:#{@port}")
+    begin
+      @server_socket = TCPServer.new(@host, @port)
+    rescue Errno::EADDRINUSE
+      @logger.error("Could not start tcp server: Address in use", host: @host, port: @port)
+      raise
+    end
+    if @ssl_enable
+      @server_socket = OpenSSL::SSL::SSLServer.new(@server_socket, @ssl_context)
+    end # @ssl_enable
+    @client_threads = Concurrent::Array.new
+
+    @accept_thread = Thread.new(@server_socket) do |server_socket|
+      LogStash::Util.set_thread_name("[#{pipeline_id}]|output|tcp|server_accept")
+      loop do
+        break if @closed.value
+        client_socket = server_socket.accept_nonblock exception: false
+        if client_socket == :wait_readable
+          IO.select [ server_socket ]
+          next
+        end
+        Thread.start(client_socket) do |client_socket|
+          # monkeypatch a 'peer' method onto the socket.
+          client_socket.instance_eval { class << self; include ::LogStash::Util::SocketPeer end }
+          @logger.debug("accepted connection", client: client_socket.peer, server: "#{@host}:#{@port}")
+          client = Client.new(client_socket, @logger)
+          Thread.current[:client] = client
+          LogStash::Util.set_thread_name("[#{pipeline_id}]|output|tcp|client_socket-#{@client_threads.size}")
+          @client_threads << Thread.current
+          client.run unless @closed.value
         end
       end
+    end
 
-      @codec.on_event do |event, payload|
-        @client_threads.select!(&:alive?)
-        @client_threads.each do |client_thread|
-          client_thread[:client].write(payload)
-        end
+    @codec.on_event do |event, payload|
+      @client_threads.select!(&:alive?)
+      @client_threads.each do |client_thread|
+        client_thread[:client].write(payload)
```

Comment:
`IO#write` used here by the server mode has the same issue as the `IO#syswrite` used in our client-mode, where it returns the number of bytes written and we need to use that feedback to determine how to send the rest of the payload. In this case though, we need to make sure we start each of the connected clients with the _whole_ payload, so binding it to a local-variable makes things more clear. 

```suggestion
        remaining_payload = payload
        while remaining_payload.bytesize > 0
          bytes_written = client_thread[:client].write(remaining_payload)
          remaining_payload = remaining_payload.byteslice(bytes_written..-1)
        end
```

This code also doesn't do anything to make sure that the client is writable, only that its thread hadn't finished before we attempt to send, so there is a race-condition where a dead client can cause an error that prevents the payload from being written to any of the _other_ clients.

Reply:
Ah `client_thread[:client].write(payload)` pushes to queue rather than directly accessing the `socket`. And `Client#run` continuously pops from queue and writes to socket where exactly this logic should be located.

Reply:
Ah. That makes sense.